src/v/cloud_storage/inventory/inv_consumer.cc

```diff
@@ -0,0 +1,259 @@
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "cloud_storage/inventory/inv_consumer.h"
+
+#include "bytes/iostream.h"
+#include "cloud_storage/logger.h"
+#include "hashing/xx.h"
+#include "serde/serde.h"
+
+#include <seastar/core/file-types.hh>
+#include <seastar/core/file.hh>
+#include <seastar/core/fstream.hh>
+#include <seastar/core/seastar.hh>
+
+#include <re2/re2.h>
+
+#include <ranges>
+
+namespace ranges = std::ranges;
+namespace views = std::views;
+
+namespace {
+// hash-string/ns/tp/partition_rev/.*
+const RE2 path_expr{"^[[:xdigit:]]+/(.*?)/(.*?)/(\\d+)_\\d+/.*?"};
+struct flush_entry {
+    model::ntp ntp;
+    fragmented_vector<uint64_t> hashes;
+    uint64_t file_name;
+};
+
+// Will remove nested, embedded quotes and spaces. The CSV report format we
+// expect should be resilient to this because we only care about path names,
+// which contain neither spaces nor quotes, but this is not a real CSV parser.
+bool drop_chars(char c) { return c != '"' && !std::isspace(c); }
+
+ss::sstring trim_string(ranges::subrange<char*> r) {
+    auto filtered = r | views::filter(drop_chars);
+    return {filtered.begin(), filtered.end()};
+}
+
+} // namespace
+
+namespace cloud_storage::inventory {
+
+inventory_consumer::inventory_consumer(
+  std::filesystem::path hash_store_path,
+  absl::node_hash_set<model::ntp> ntps,
+  size_t max_paths_in_memory)
+  : _hash_store_path{std::move(hash_store_path)}
+  , _ntps{std::move(ntps)}
+  , _max_hash_size_in_memory{max_paths_in_memory} {}
+
+ss::future<> inventory_consumer::consume(
+  ss::input_stream<char> s, is_gzip_compressed is_report_compressed) {
+    auto h = _gate.hold();
+    report_parser parser{std::move(s), is_report_compressed};
+    co_await parser
+      .consume([this](auto&& paths) { return process_paths(std::move(paths)); })
+      .finally([&parser] { return parser.stop(); });
+    co_await flush(write_all_t::yes);
+}
+
+std::vector<ss::sstring> inventory_consumer::parse_row(ss::sstring row) {
+    std::vector<ss::sstring> pieces;
+    for (auto p : row | views::split(',') | views::transform(trim_string)) {
+        pieces.emplace_back(p);
+    }
+    return pieces;
+}
+
+ss::future<>
+inventory_consumer::process_paths(fragmented_vector<ss::sstring> paths) {
+    for (const auto& path : paths) {
+        // The row is expected to be in the form:
+        // "bucket","path"
+        auto pieces = parse_row(path);
+        // If a row is malformed we skip it and do not abort processing.
+        // This way the rest of the entries are processed, and if there
+        // was a path in this row it will be recorded as missing.
+        if (pieces.size() != 2) {
+            vlog(cst_log.warn, "unexpected row in inventory report: {}", path);
+        }
```

Comment:
Are we guaranteed that there is only one bucket? Should we filter on bucket?

Reply:
For a given report we are guaranteed to have just one bucket https://docs.aws.amazon.com/AmazonS3/latest/userguide/storage-inventory.html#storage-inventory-buckets